### PR TITLE
Protected IPFS api

### PR DESCRIPTION
**Pull request by yornaath**
Implement a protected IPFS api available to the app that validates metadata and its size.

**Note:**
When this is merged to mainnet app and goes live the cluster authentication username and password should be updated and updated in the env variables and then the app redeployed:

### CODE REVIEW

File: pages/api/ipfs/index.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
+import { IOMarketMetadata } from "@zeitgeistpm/sdk";
+import type { PageConfig } from "next";
+import { NextRequest } from "next/server";
+
+const CLUSTER_ENDPOINT = process.env.IPFS_CLUSTER_URL;
+
+const headers = {
+  Authorization: `Basic ${Buffer.from(
+    `${process.env.IPFS_CLUSTER_USERNAME}:${process.env.IPFS_CLUSTER_PASSWORD}`,
```
Comment (sea212) [inline pages/api/ipfs/index.ts:9]: Those will never be exposed to the end user?

File: pages/api/ipfs/index.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  ).toString("base64")}`,
+};
+
+export const config: PageConfig = {
+  runtime: "edge",
+};
+
+export default async function handler(req: NextRequest) {
+  if (req.method === "POST") {
+    const [error, parsed] = IOMarketMetadata.validate(
```
Comment (sea212) [inline pages/api/ipfs/index.ts:19]: It ensures that the form of the json matches what we expect as valid market metadata?

File: lib/metadata-storage.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import { u8aConcat } from "@polkadot/util/u8a";
+import { MetadataStorage, createStorage } from "@zeitgeistpm/sdk";
+import { Codec, JsonCodec } from "@zeitgeistpm/utility/dist/codec";
+import * as O from "@zeitgeistpm/utility/dist/option";
+import * as Te from "@zeitgeistpm/utility/dist/taskeither";
+import { Storage, StorageError } from "@zeitgeistpm/web3.storage";
+import * as IPFSHTTPClient from "ipfs-http-client";
+import { CID } from "multiformats/cid";
+
+const node = IPFSHTTPClient.create({ url: "https://ipfs.zeitgeist.pm" });
```
Comment (sea212) [inline lib/metadata-storage.ts:11]: Why is that hardcoded?

File: lib/metadata-storage.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+export const createMetadataStorage = (): MetadataStorage => {
```
Comment (sea212) [inline lib/metadata-storage.ts:13]: When is this expected to be used and when is the cluster expected to be used?

File: pages/api/ipfs/index.ts (hunk 1 above)
Comment (yornaath) [reply pages/api/ipfs/index.ts:9]: No those are only loaded on the server. Also vercel/next prevents any env varible that isnt prefixed with `NEXT_PUBLIC` from being sent to the client.

File: pages/api/ipfs/index.ts (hunk 2 above)
Comment (yornaath) [reply pages/api/ipfs/index.ts:19]: Yes

File: lib/metadata-storage.ts (hunk 3 above)
Comment (yornaath) [reply lib/metadata-storage.ts:11]: Doesnt need to be.

File: lib/metadata-storage.ts (hunk 4 above)
Comment (yornaath) [reply lib/metadata-storage.ts:13]: This is only used to read data from IPFS so this should be set to the new hosted ipfs node I believe. Cluster is used to add/pin files and remove them when rolling back a failed transaction.

File: lib/metadata-storage.ts (hunk 3 above)
Comment (sea212) [reply lib/metadata-storage.ts]: Hmm I think it is sufficient though, don't see why we would want to change this.

File: lib/metadata-storage.ts (hunk 4 above)
Comment (sea212) [reply lib/metadata-storage.ts:15]: We will shortly shut down the cluster and only use a singular hosted IPFS node (replication happens also in Subsquid). For now this is the right approach though, let's adjust once we have the hosted IPFS node ready.

**Review by sea212 — APPROVED** [review]

File: pages/api/ipfs/types.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import * as z from "zod";
+
+export const IOMarketMetadata = z.object({
```
Comment (Robiquet) [inline pages/api/ipfs/types.ts:3]: Might be nice to share with market form validation and add some rough size restrictions if possible

**Review by Robiquet — APPROVED** [review]

**Pull request merged**